scalalib/src/mill/scalalib/SemanticDbJavaModule.scala

```diff
@@ -72,45 +99,91 @@ trait SemanticDbJavaModule extends JavaModule { hostModule =>
+    }
+    def cleanedClassesDir(classesDir: os.Path): PathRef = {
+      os.walk(classesDir, preOrder = true)
+        .filter(os.isFile)
+        .foreach { p =>
+          if (p.ext != "semanticdb") {
+            os.remove(p)
+          }
+        }
+      PathRef(classesDir)
+    }
+
+    hostModule match {
+      // TODO: Avoid fetching the Java semanticdb version when there are no Java sources (better detect mixed)
+      case m: ScalaModule =>
+        T.persistent {
+          val sv = m.scalaVersion()
+
+          val scalacOptions = (
+            m.allScalacOptions() ++
+              semanticDbEnablePluginScalacOptions() ++ {
+                if (ZincWorkerUtil.isScala3(sv)) {
+                  Seq("-Xsemanticdb")
+                } else {
+                  Seq(
+                    "-Yrangepos",
+                    s"-P:semanticdb:sourceroot:${T.workspace}",
+                    "-Ystop-after:semanticdb-typer"
+                  )
+                }
               }
-            }
-        )
-          .filterNot(_ == "-Xfatal-warnings")
-        T.log.debug(s"effective scalac options: ${scalacOptions}")
-
-        zincWorker
-          .worker()
-          .compileMixed(
-            upstreamCompileOutput = upstreamCompileOutput(),
-            sources = m.allSourceFiles().map(_.path),
-            compileClasspath = compileClasspath().map(_.path),
-            javacOptions = javacOptions(),
-            scalaVersion = sv,
-            scalaOrganization = m.scalaOrganization(),
-            scalacOptions = scalacOptions,
-            compilerClasspath = m.scalaCompilerClasspath(),
-            scalacPluginClasspath = semanticDbPluginClasspath(),
-            reporter = T.reporter.apply(hashCode)
           )
-          .map(_.classes)
-      }
+            .filterNot(_ == "-Xfatal-warnings")
+          T.log.debug(s"effective scalac options: ${scalacOptions}")
+
+          zincWorker.worker()
+            .compileMixed(
+              upstreamCompileOutput = upstreamCompileOutput(),
+              sources = m.allSourceFiles().map(_.path),
+              compileClasspath = compileClasspathTask(m)().map(_.path),
+              javacOptions = javacOptionsTask(m)(),
+              scalaVersion = sv,
+              scalaOrganization = m.scalaOrganization(),
+              scalacOptions = scalacOptions,
+              compilerClasspath = m.scalaCompilerClasspath(),
+              scalacPluginClasspath = semanticDbPluginClasspath(),
+              reporter = T.reporter.apply(hashCode)
+            )
+            .map(r => cleanedClassesDir(r.classes.path))
+        }
+      case m: JavaModule =>
+        T.persistent {
+          val javacOpts = javacOptionsTask(m)()
+
+          // we currently assume, we don't do incremental java compilation
+          os.remove.all(T.dest / "classes")
+//          val semDbPath = T.dest / "_semanticdb"
+//          os.remove.all(semDbPath)
```

Comment:
Not sure if these are overlooked or not, but if we don't need them we should probably just remove them.

Reply:
Yeah, I'm still editing, so I try to spare extra commits and CI cycles.